src/buildstream/sandbox/sandbox.py

```diff
@@ -526,6 +526,32 @@ def _create_empty_file(self, name):
             cwd_vdir = vdir.open_directory(cwd.lstrip(os.sep), create=True)
             cwd_vdir._create_empty_file(name)
 
+    # _clean_build_directory()
```

Comment:
I'm not particularly fond of the business logic growing here in the sandbox code, it looks like we're building on this `Sandbox._set_build_directory()` API which appears to be deadcode which we neglected to cleanup (grep tells me that there is no code observing this `_build_directory` variable anywhere).

Maybe it would make more sense to nuke these deadcode paths and some more generic private cleanup API controlled by element.py's business logic decisions.

Reply:
Would it be acceptable to check `context.cache_buildtrees` in `BuildElement` as condition to calling a more generic method in `Sandbox`?

Or should there be a (public?) `clean_build_directory()` method in `Element`, which performs that check and calls the `Sandbox` method?

Reply:
> Would it be acceptable to check `context.cache_buildtrees` in `BuildElement` as condition to calling a more generic method in `Sandbox`?

Looking at `BuildElement` now, it looks like we've done a good job of avoiding any internal API interactions - it would be a good thing to keep it this way (anything a build element can do, any external plugin can do).

> Or should there be a (public?) clean_build_directory() method in Element, which performs that check and calls the Sandbox method?

Is there any reason why this cleaning of the build directory should be optional ?

Perhaps a better criteria would be to clean the build directory in any case where `%{build-root}` is found outside of `%{install-root}` (this would avoid destructive behavior like in script elements where `%{build-root}` might be inside, or equal to the `%{install-root}`) ?

Reply:
If we clean the build directory even with the `cache-buildtrees: always` configuration, that option becomes useless, if I'm not missing anything.

Reply:
I've moved the business logic to a `Element.clean_buildtree()` method. Is this what you had in mind?

Reply:
> If we clean the build directory even with the `cache-buildtrees: always` configuration, that option becomes useless, if I'm not missing anything.

Maybe I'm missing something but I thought the sanbox functions in question were already only *maybe_clean_buildtree* depending on some business logic.

Reply:
After discussing on IRC, @juergbi reminds me that this function must run commands in a shell in order to be batched and avoid capturing, which sort of explains why this must be explicitly invoked by the plugin.

* It must happen at the end of a batch in order to preserve the optimization and avoid the unnecessary capture
* It must only be performed by an element which has a shell capable or running `rm`

As such, this approach appears to make sense.

We should make it obvious in the function name or docstrings that this cleanup requires a shell to work.